internal/stores/contracts.go

```diff
@@ -179,3 +194,382 @@ func NewJSONContractStore(dir string) (*JSONContractStore, error) {
 	}
 	return s, nil
 }
+
+type (
+	// SQLContractStore implements the bus.ContractStore interface using SQL as the
+	// persistence backend.
+	SQLContractStore struct {
+		db *gorm.DB
+	}
+
+	dbContractRHPv2 struct {
+		ID         types.FileContractID     `gorm:"primaryKey,type:bytes;serializer:gob;NOT NULL"`
+		Revision   dbFileContractRevision   `gorm:"constraint:OnDelete:CASCADE;foreignKey:ParentID;references:ID;NOT NULL"` //CASCADE to delete revision too
+		Signatures []dbTransactionSignature `gorm:"constraint:OnDelete:CASCADE;foreignKey:ParentID;references:ID;NOT NULL"` // CASCADE to delete signatures too
+	}
+
+	dbFileContractRevision struct {
```

Comment:
in my head the revisions do indeed have their own database table but they would reference the contract and have a  raw field that contains the entire revision as opposed to adding separate entities for it, do we have a lot to gain by splitting it out fully?

Reply:
oh ok after reading the entire PR I noticed `gorm` takes care of everything on insert, decided to leave the comment because we might want to consider implementing a [custom data type instead ](https://gorm.io/docs/data_types.html#Implements-Customized-Data-Type)

Reply:
Yeah gorm does a ton in the background. Every struct is its own table and slices are essentially relations between them. I only chose a raw type for the `CoveredFields` field cause there is no point in having that thing turned into a table with columns but other things like startHeights, endHeights, sizes, revision numbers and so on might be useful to have as columns to query them.

I did however simplify it a bit to reduce the number of tables for things that we probably never want to query for.

Reply:
great, thanks